engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java

```diff
@@ -21,47 +23,71 @@
 @ExportLibrary(WarningsLibrary.class)
 @ExportLibrary(ReflectionLibrary.class)
 public final class WithWarnings implements TruffleObject {
+
   private final EconomicSet<Warning> warnings;
   private final Object value;
 
-  private WithWarnings(Object value, Warning... warnings) {
+  private final boolean reachedMax;
+  private final int maxWarnings;
+
+  private WithWarnings(Object value, int maxWarnings, boolean reachedMaxCount, Warning... warnings) {
     assert !(value instanceof WithWarnings);
-    this.warnings = createSetFromArray(warnings);
+    this.warnings = createSetFromArray(maxWarnings, warnings);
     this.value = value;
+    this.reachedMax = reachedMaxCount || this.warnings.size() >= maxWarnings;
+    this.maxWarnings = maxWarnings;
+  }
+  private WithWarnings(Object value, int maxWarnings, Warning... warnings) {
+    this(value, maxWarnings, false, warnings);
   }
 
-  private WithWarnings(Object value, EconomicSet<Warning> warnings, Warning... additionalWarnings) {
+  private WithWarnings(Object value, int maxWarnings, EconomicSet<Warning> warnings, boolean reachedMaxCount, Warning... additionalWarnings) {
```

Comment:
It looks concerning to me that checking if we reached `reachedMaxCount` is the responsibility of the caller. It should be responsibility of _this_ class to ensure this is correctly computed. I can imagine very easily the caller messing this up.

If you could please:
1. make this computation opaque to the caller - its an implementation detail of the WithWarnings
2. OR at least please document what this argument means precisely and when it needs to be set.

Reply:
This is intentional due to different limits that different methods can have. For now this only concerns `Vector.sort`. 

I will add a comment.

Reply:
Hmm, I thought that the "Additional warnings" message added in the protocol should only concern the _global_ limit. Making it depend on different limits will make stuff confusing IMO.

As for the different limit for Vector - maybe there we should just construct a 'Additional Warnings' instance containing the count of suppressed comparison warnings?

Reply:
IMHO the vector limit should be done at 'user level' - i.e. not touch the global warning count limiter at all - instead here we can count how many warnings were suppressed and have a special warning type that will report 'and X suppressed warnings'.

But if it's a problem to implement this, then maybe let's just not have a separate limit after all? To at least keep the global limiting logic simple. Currently it looks quite complex and I cannot be sure that we will not have errors when computing these flags.